pydm/widgets/archiver_time_plot.py

```diff
@@ -286,6 +292,313 @@ def receiveNewValue(self, new_value):
+        """
+        if not self.checkFormula():
+            self.formula_invalid_signal.emit()
+            return
+
+        pvArchiveData = dict()
+        pvLiveData = dict()
+        pvIndices = dict()
+        pvValues = dict()
+        self.minx = float("-inf")
+        self.maxx = float("inf")
+        formula = self._trueFormula
+        self.archive_data_buffer = np.zeros((2, 0), order="f", dtype=float)
+        self.data_buffer = np.zeros((2, 0), order="f", dtype=float)
+        # Reset buffers
+        self.points_accumulated = 0
+        self.archive_points_accumulated = 0
+        # Populate new dictionaries, simply for ease of access and readability
+        for pv in self.pvs.keys():
+            pvArchiveData[pv] = self.pvs[pv].archive_data_buffer
+            pvIndices[pv] = 0
+            # Of all the rows we are relying on, we want our min to be the largest of their mins. Opposite for maxes.
+            # Only want to attempt to draw the curve where we have all required data for it.
+            self.minx = max(self.pvs[pv].min_archiver_x(), self.minx)
+            self.maxx = min(self.pvs[pv].max_archiver_x(), self.maxx)
+        for pv in self.pvs.keys():
+            pvValues[pv] = pvArchiveData[pv][1][pvIndices[pv]]
+            while pvIndices[pv] < len(pvArchiveData[pv][0]) - 1 and pvArchiveData[pv][0][pvIndices[pv]] < self.minx:
+                pvValues[pv] = pvArchiveData[pv][1][pvIndices[pv]]
+                pvIndices[pv] += 1
+                # Shift starting indices for each row to our minimum
+        x = self.minx
+        while True:
+            self.archive_points_accumulated += 1
+            minPV = None
+            # Find the next x point out of all of our rows.
+            # Update only that row's value, use the previous value of other rows for calcs.
+            for pv in self.pvs.keys():
+                if minPV is None or pvArchiveData[pv][0][pvIndices[pv]] < pvArchiveData[minPV][0][pvIndices[minPV]]:
+                    minPV = pv
+                    x = pvArchiveData[pv][0][pvIndices[pv]]
+
+            pvValues[minPV] = pvArchiveData[minPV][1][pvIndices[minPV]]
+            try:
+                temp = np.array([[x], [eval(formula)]])
+            except ValueError:
+                print("Evaluate failed (domain errors? unknown function?)")
+                temp = np.array([[x], [0]])
+            self.archive_data_buffer = np.append(self.archive_data_buffer, temp, axis=1)
+            pvIndices[minPV] += 1
+            # If we are out of data for this row, stop!
+            if pvIndices[minPV] >= len(pvArchiveData[minPV][0]):
+                break
+        if self.liveData:
+            formula = formula.replace("Archive", "Live")
+            minx = float("-inf")
+            maxx = float("inf")
+            self.points_accumulated = 0
+            pvIndices = dict()
+            pvValues = dict()
+            # Do literally the exact same thing for live data
```

Comment:
since it looks like we're doing the same logic for live and archive but just using different data, 
could we check if using live/archive and then set a common variable like 'pvData' to either live or archive set depending?

for example this chunk:
```
        for pv in self.pvs.keys():
            pv_archive_times = pvArchiveData[pv][0]
            pv_archive_values = pvArchiveData[pv][1]
            pv_current_index = pvIndices[pv]
            pvValues[pv] = pv_archive_values[pv_current_index]
            while pv_current_index < len(pv_archive_times) - 1 and pv_archive_times[pv_current_index] < self.minx:
                pvValues[pv] = pv_archive_values[pv_current_index]
                pv_current_index += 1
                # Shift starting indices for each row to our minimum
            pvIndices[pv] = pv_current_index
````
is the same as this chunk:
````
            for pv in self.pvs.keys():
                pv_live_times = pvLiveData[pv][0]
                pv_live_values = pvLiveData[pv][1]
                pv_current_index = pvIndices[pv]
                pvValues[pv] = pv_live_values[pv_current_index]
                while pv_current_index < len(pv_live_times) - 1 and pv_live_times[pv_current_index] < minx:
                    pvValues[pv] = pv_live_values[pv_current_index]
                    pv_current_index += 1

                pvIndices[pv] = pv_current_index
````


except for which map we are acting on. Could we just use a function arg to set some variable like 'pvData' to replace 'pvLiveData' and 'pvArchieveData'

Reply:
Right, I understand. Done now. Thanks!

Reply:
Right, now I understand. Changed, thanks!